### PR TITLE
[Core] Register Project handlers in FileWatcherService

### PR DESCRIPTION
**Pull request by Therzok**
Second attempt at fixing filtering of notifications when using our big cone file system watcher.

See API usage in test files.

Fixes VSTS #643326 - PathTree should be able to remove a node directly
Fixes VSTS #947787 - Design API for consuming filtered FileService+FileWatcher notifications
Fixes VSTS #947789 - Use filtered FileService+FileWatcher APIs in slow handlers
Fixes VSTS #947788 - Implement API design for FileService+FileWatcher filtered notifications

This should fix our problems with projects registering on global events and only be notified on items changed in their subpaths.

### CODE REVIEW

**Comment by Therzok** [comment]
We will fix some problems in a future iteration:

* Inversion of control, so Project registers a handler in the PathTree
* FileChanged event - that requires passing a fixed set of files
* Non-recursive directory watching

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core/FileService.cs [hunk 1]
```diff
@@ -758,8 +758,6 @@ public static void ThawEvents ()
 		public static event EventHandler<FileEventArgs> FileCreated;
 		static void OnFileCreated (FileEventArgs args)
 		{
-			AsyncEvents.OnFileCreated (args);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core/FileService.cs]: Pretty sure nothing uses AsyncEvents, so it was safe to remove.

File: main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/FileWatcherTests.cs [hunk 2]
```diff
@@ -38,134 +38,11 @@
 namespace MonoDevelop.Projects
 {
 	[TestFixture]
-	public class FileWatcherTests : TestBase
+	public class FileWatcherTests : FileWatcherTestBase
 	{
 		Solution sol;
-		List<FileEventInfo> fileChanges;
```
Comment (Therzok) [inline main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/FileWatcherTests.cs]: All these were moved to the new base class.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 3]
```diff
@@ -4561,35 +4537,15 @@ void OnFileCreated (FilePath filePath)
...
+		internal virtual void OnFileDeleted (FilePath filePath)
 		{
-			if (Runtime.IsMainThread)
+			if (!eventsEnabled)
 				return;
 
-			List<FilePath> paths = null;
-			foreach (var file in e) {
-				var path = file.FileName;
-				bool exists = File.Exists (path) || Directory.Exists (path);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: Already done [here](https://github.com/mono/monodevelop/pull/8273/files#diff-66c041c389296698df1952f008e6beb5R401)

**Comment by Therzok** [comment]
```
2019-07-25T15:56:42.4264380Z 1) WatchDirectories_TwoFilesChanged_OneClosed (MonoDevelop.Projects.FileWatcherTests.WatchDirectories_TwoFilesChanged_OneClosed)
2019-07-25T15:56:42.4264710Z Expected: False
2019-07-25T15:56:42.4264840Z But was: True
2019-07-25T15:56:42.4264890Z
2019-07-25T15:56:42.4265070Z at MonoDevelop.Projects.FileWatcherTests.WatchDirectories_TwoFilesChanged_OneClosed () [0x003a0] in /Users/vsts/agent/2.154.3/work/1/s/monodevelop/main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/FileWatcherTests.cs:399
2019-07-25T15:56:42.4271190Z at MonoDevelop.Projects.FileWatcherTests.WatchDirectories_TwoFilesChanged_OneClosed () [0x00433] in /Users/vsts/agent/2.154.3/work/1/s/monodevelop/main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/FileWatcherTests.cs:401
```

Tried running locally, cannot see that. Maybe I missed a file watcher unregistration?

**Comment by Therzok** [comment]
Also, PathTree added new leak. Will take a look.

**Comment by Therzok** [comment]
Should go green now 🎉

**Comment by Therzok** [comment]
@slluis ping

**Review by mrward — APPROVED** [review]

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs [hunk 4]
```diff
@@ -265,8 +382,20 @@ static void OnFileChanged (object sender, FileSystemEventArgs e)
...
 		}
 
-		static void OnFileCreated (object sender, FileSystemEventArgs e)
+		void OnFileCreated (object sender, FileSystemEventArgs e)
 		{
+			using (readerWriterLock.Read ()) {
+				var sw = FileWatcherService.Timings.Get ();
+				sw.Restart ();
+				NotifyNode (rootNode, e.FullPath, (id, path) => {
+					if (id is Project project) {
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: This should not be harcoded to Project. A better solution would be to define an interface with OnFileCreated, OnFileDeleted, etc. methods.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs [hunk 5]
```diff
@@ -265,8 +382,20 @@ static void OnFileChanged (object sender, FileSystemEventArgs e)
...
+						project.OnFileCreated (path);
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: Shouldn't this be called in the UI thread?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs (hunk 4 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: Yes, I wanted to do so initially, but can we do that in another iteration when we port over other types?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs (hunk 5 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: Used to be on AsyncEvents. Cc @mrward

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: I sent an incomplete comment. AsyncEvents used to be called on a background thread, so the behaviour is unchanged for Project.

Comment (slluis) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: 👍

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs (hunk 4 above)
Comment (slluis) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: OK

**Review by slluis — APPROVED** [review]

**Pull request merged**